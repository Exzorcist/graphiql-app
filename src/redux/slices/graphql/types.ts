import { IntrospectionQuery } from 'graphql';
import { AsyncStatus } from '@/types/AsyncStatus';

export type IntrospectionState = {
  readonly data: IntrospectionQuery | null;
  endpoint: string;
  status: AsyncStatus;
};
