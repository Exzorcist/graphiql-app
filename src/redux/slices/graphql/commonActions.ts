import { createAction } from '@reduxjs/toolkit';
import { GraphQLSchema } from 'graphql';
import { IntrospectionState } from './types';

function withPayloadType<T>() {
  return (t: T) => ({ payload: t });
}

export const setGraphqlSchema = createAction(
  'graphql/setGraphqlSchema',
  withPayloadType<GraphQLSchema>()
);

export const setIntrospection = createAction(
  'graphql/setIntrospection',
  withPayloadType<Partial<IntrospectionState>>()
);
