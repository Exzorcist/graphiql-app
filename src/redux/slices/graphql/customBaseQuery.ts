import {
  BaseQueryFn,
  FetchArgs,
  FetchBaseQueryError,
  FetchBaseQueryMeta,
  fetchBaseQuery,
} from '@reduxjs/toolkit/query';

export type Meta = {
  responseTime: number;
};

export type CustomBaseQueryMeta = Meta & FetchBaseQueryMeta;

const fetchWithBQ = fetchBaseQuery({ method: 'POST' });

export const customBaseQuery: BaseQueryFn<
  FetchArgs,
  unknown,
  FetchBaseQueryError,
  NonNullable<unknown>,
  CustomBaseQueryMeta
> = async (args, api, extraOptions) => {
  const timestampBefore = Date.now();
  const baseResult = await fetchWithBQ(args, api, extraOptions);
  const responseTime = Date.now() - timestampBefore;

  return {
    ...baseResult,
    meta: baseResult.meta && { ...baseResult.meta, responseTime },
  };
};
