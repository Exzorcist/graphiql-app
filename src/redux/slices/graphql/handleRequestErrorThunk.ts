import { AppDispatch, RootState } from '@/redux/store';
import { isFetchBaseQueryError, isErrorWithMessage } from '@/utils/type-guards';
import { setMessage } from '../globalMessageSlice';
import { getFetchErrorMsg } from './utils';

export const handleFetchErrorThunk =
  (error: unknown) => (dispatch: AppDispatch, getState: () => RootState) => {
    if (isFetchBaseQueryError(error)) {
      dispatch(
        setMessage({
          type: 'error',
          isShown: true,
          text: getFetchErrorMsg(error, getState().localization),
        })
      );
    } else if (isErrorWithMessage(error)) {
      dispatch(setMessage({ type: 'error', isShown: true, text: error.message }));
    }
  };
