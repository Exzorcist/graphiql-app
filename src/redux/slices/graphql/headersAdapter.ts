import { createEntityAdapter, createSelector } from '@reduxjs/toolkit';
import { HttpHeader } from '@/types/HttpHeader';
import { RootState } from '@/redux/store';

export const headersAdapter = createEntityAdapter<HttpHeader>();
export const headersInitialState = headersAdapter.getInitialState();

export const { selectAll: selectHeaders } = headersAdapter.getSelectors(
  (state: RootState) => state.graphql.headers
);

export const selectHeaderMap = createSelector(selectHeaders, (headers) => {
  return headers.reduce((acc, { header: headerKey, value: headerValue }) => {
    const key = headerKey.trim();
    const value = headerValue.trim();

    if (!key || !value) return acc;

    return { ...acc, [key]: value };
  }, {});
});
