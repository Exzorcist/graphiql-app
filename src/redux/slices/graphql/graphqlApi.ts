import { IntrospectionQuery, buildClientSchema, getIntrospectionQuery } from 'graphql';
import { createApi } from '@reduxjs/toolkit/query/react';
import { setGraphqlSchema, setIntrospection } from './commonActions';
import { handleFetchErrorThunk } from './handleRequestErrorThunk';
import { emptySchema } from '@/utils/emptyGraphqlSchema';
import { customBaseQuery } from './customBaseQuery';
import { setMessage } from '../globalMessageSlice';
import { getSchemaReloadMsg } from './utils';
import { RootState } from '@/redux/store';

export const graphqlApi = createApi({
  reducerPath: 'graphqlApi',
  baseQuery: customBaseQuery,
  endpoints: (builder) => ({
    initRequest: builder.mutation<unknown, string | void>({
      queryFn: async (operationName, { getState, dispatch }, _extraOptions, fetchWithBQ) => {
        const state = getState() as RootState;
        const { graphql } = state;

        const response = await fetchWithBQ({
          url: graphql.endpointValue,
          body: { query: graphql.request.value, variables: graphql.variablesValue, operationName },
        });

        const err = response.error;

        if (err) {
          if (err.data) {
            return { ...response, data: err.data, error: undefined };
          }

          dispatch(handleFetchErrorThunk(err));
        }

        return response;
      },
    }),

    fetchIntrospection: builder.mutation<IntrospectionQuery, string>({
      query: (url) => ({
        url,
        body: { query: getIntrospectionQuery() },
      }),
      transformResponse: (res: { data: IntrospectionQuery }) => {
        return res.data;
      },
      onQueryStarted: async (arg, { getState, queryFulfilled, dispatch }) => {
        dispatch(setIntrospection({ status: 'pending' }));
        const { localization } = getState() as RootState;

        try {
          const resp = await queryFulfilled;
          dispatch(setGraphqlSchema(buildClientSchema(resp.data)));
          dispatch(setIntrospection({ data: resp.data, endpoint: arg, status: 'fullfilled' }));
          dispatch(
            setMessage({ text: getSchemaReloadMsg(localization), isShown: true, type: 'success' })
          );
        } catch (err) {
          dispatch(setGraphqlSchema(emptySchema));
          dispatch(setIntrospection({ data: null, endpoint: '', status: 'rejected' }));

          if (err && typeof err === 'object' && 'error' in err) {
            dispatch(handleFetchErrorThunk(err.error));
          }
        }
      },
    }),
  }),
});

export const { useFetchIntrospectionMutation, useInitRequestMutation } = graphqlApi;
