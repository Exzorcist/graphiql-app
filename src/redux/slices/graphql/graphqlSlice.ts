import { buildClientSchema, GraphQLSchema } from 'graphql';
import { Draft, PayloadAction, createSlice } from '@reduxjs/toolkit';
import { REHYDRATE } from 'redux-persist';
import { emptySchema } from '@/utils/emptyGraphqlSchema';
import { CustomBaseQueryMeta } from './customBaseQuery';
import { AsyncStatus } from '@/types/AsyncStatus';
import { graphqlApi } from './graphqlApi';
import { IntrospectionState } from './types';

export type GraphqlSliceState = {
  request: { value: string; status: AsyncStatus };
  hasRequestEditorLintErrors: boolean;
  introspection: IntrospectionState;
  schema: GraphQLSchema;
  response: {
    data: unknown;
    statusCode?: number;
    responseTime?: number;
  } | null;
  endpointValue: string;
  variablesValue: object;
};

export const initialState: GraphqlSliceState = {
  introspection: { data: null, endpoint: '', status: 'idle' },
  request: { value: '', status: 'idle' },
  hasRequestEditorLintErrors: false,
  variablesValue: {},
  endpointValue: '',
  response: null,
  schema: emptySchema,
};

export const graphqlSlice = createSlice({
  name: 'graphql',
  initialState,
  reducers: {
    changeRequestValue(state, action: PayloadAction<string>) {
      state.request.value = action.payload;
    },
    changeEndpointValue(state, action: PayloadAction<string>) {
      state.endpointValue = action.payload;
    },
    changeVariablesValue(state, action: PayloadAction<object>) {
      state.variablesValue = action.payload;
    },
    setHasRequestEditorLintErrors(state, action: PayloadAction<boolean>) {
      state.hasRequestEditorLintErrors = action.payload;
    },
    setGraphqlSchema(state, action: PayloadAction<GraphQLSchema>) {
      return { ...state, schema: action.payload };
    },
    setIntrospection(state, action: PayloadAction<Partial<IntrospectionState>>) {
      state.introspection = {
        ...state.introspection,
        ...action.payload,
      } as Draft<IntrospectionState>;
    },
  },

  extraReducers(builder) {
    builder
      .addCase(REHYDRATE, (state, _action) => {
        const action = _action as PayloadAction<GraphqlSliceState | undefined> & { key: string };

        if (action.key === 'graphql' && action.payload?.introspection.data) {
          try {
            const schema = buildClientSchema(action.payload.introspection.data);
            return { ...state, schema };
          } catch (error) {
            return {
              ...state,
              introspection: initialState.introspection as Draft<IntrospectionState>,
            };
          }
        }

        return state;
      })
      .addMatcher(graphqlApi.endpoints.initRequest.matchFulfilled, (state, action) => {
        const meta = action.meta.baseQueryMeta as CustomBaseQueryMeta | undefined;
        const responseTime = meta?.responseTime;
        const statusCode = meta?.response?.status;

        state.request.status = 'fullfilled';
        state.response = { data: action.payload, statusCode, responseTime };
      })
      .addMatcher(graphqlApi.endpoints.initRequest.matchPending, (state) => {
        state.request.status = 'pending';
      })
      .addMatcher(graphqlApi.endpoints.initRequest.matchRejected, (state) => {
        state.request.status = 'rejected';
      });
  },

  selectors: {
    selectResponse: (state) => state.response,
    selectGraphQLSchema: (state) => state.schema,
    selectGraphQLDocsSchema: (state) => (state.schema === emptySchema ? null : state.schema),
    selectRequestValue: (state) => state.request.value,
    selectEndpointValue: (state) => state.endpointValue,
    selectRequestStatus: (state) => state.request.status,
    selectResponseValue: (state) => state.response?.data,
    selectVariablesValue: (state) => state.variablesValue,
    selectIntrospectStatus: (state) => state.introspection.status,
    selectIntrospectEndpoint: (state) => state.introspection.endpoint,
    selectHasRequestEditorLintErrors: (state) => state.hasRequestEditorLintErrors,
  },
});

export const graphqlReducer = graphqlSlice.reducer;

export const {
  changeRequestValue,
  changeEndpointValue,
  changeVariablesValue,
  setHasRequestEditorLintErrors,
} = graphqlSlice.actions;

export const {
  selectGraphQLDocsSchema,
  selectHasRequestEditorLintErrors,
  selectGraphQLSchema,
  selectRequestStatus,
  selectEndpointValue,
  selectRequestValue,
  selectResponse,
  selectResponseValue,
  selectVariablesValue,
  selectIntrospectStatus,
  selectIntrospectEndpoint,
} = graphqlSlice.selectors;
