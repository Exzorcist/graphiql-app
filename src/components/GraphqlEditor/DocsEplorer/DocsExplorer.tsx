import { createContext, useContext, useMemo, useRef, useState } from 'react';
import { selectGraphQLSchema, selectIntrospectStatus } from '@/redux/slices/graphql/graphqlSlice';
import { useAppSelector } from '@/utils/hooks/redux-hooks';
import RootTypeList from './components/RootTypeList';
import { GraphQLDocsEntry } from '@/types/graphqlTypes';
import Separator from '@/components/ui/Separator';
import { cn } from '@/utils/cn';
import EntryScreen from './components/EntryScreen';
import { PropsWithClassName } from '@/types/PropsWithClassName';
import DocsBreadcrumb from './components/DocsBreadcrumb';
import Spinner from '@/components/ui/Spinner';
import { useLocalizationContext } from '@/providers/LocalizationProvider';

type DocsExlorerContextType = {
  openEntry(entry: GraphQLDocsEntry): void;
  leaveEntry(): void;
  goToIndex(index: number): void;
};

const DocsExplorerContext = createContext<DocsExlorerContextType | null>(null);

function DocsExplorer({ className }: PropsWithClassName) {
  const { t } = useLocalizationContext();
  const graphqlSchema = useAppSelector(selectGraphQLSchema);
  const introspectStatus = useAppSelector(selectIntrospectStatus);
  const [navStack, setNavStack] = useState<GraphQLDocsEntry[]>([]);
  const prevSchema = useRef(graphqlSchema);

  if (prevSchema.current !== graphqlSchema) {
    prevSchema.current = graphqlSchema;

    if (navStack.length) {
      setNavStack([]);
    }
  }

  const contextValue: DocsExlorerContextType = useMemo(() => {
    return {
      openEntry(entry) {
        setNavStack((prev) => [...prev, entry]);
      },
      leaveEntry() {
        setNavStack((prev) => prev.slice(0, -1));
      },
      goToIndex(index: number) {
        setNavStack((prev) => prev.slice(0, index + 1));
      },
    };
  }, []);

  return (
    <DocsExplorerContext.Provider value={contextValue}>
      <div className="h-full flex flex-col relative">
        <div
          className={cn(
            'bg-editor-primary p-5 h-full w-full font-sans min-h-0 flex-[1_1_0px] overflow-auto fancy-scrollbar',
            introspectStatus === 'pending' && 'opacity-50 pointer-events-none',
            className
          )}
        >
          <h3 className="text-lg font-semibold">{t.page.editor.docs}</h3>
          <Separator className="mb-4" />
          <DocsBreadcrumb
            navStack={navStack}
            onItemClick={contextValue.goToIndex}
            className="mb-4"
          />
          {navStack.length ? <EntryScreen entry={navStack.at(-1)!} /> : <RootTypeList />}
        </div>
        {introspectStatus === 'pending' && <Spinner className="absolute-center" withLabel />}
      </div>
    </DocsExplorerContext.Provider>
  );
}

export function useDocsExplorer() {
  const value = useContext(DocsExplorerContext);

  if (!value) {
    throw new Error('useDocsExplorer must be used inside DocsExlorerContext');
  }

  return value;
}

export default DocsExplorer;
