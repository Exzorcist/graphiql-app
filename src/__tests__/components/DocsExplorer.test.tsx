import { IntrospectionQuery } from 'graphql';
import { screen } from '@testing-library/react';
import DocsExplorer from '@/components/GraphqlEditor/DocsEplorer/DocsExplorer';
import introspectionResponse from '../fixtures/introspectionResponse.json';
import { customRender as render, user } from '../test-utils';

test('DocsEplorer', async () => {
  render(<DocsExplorer />, {
    preloadedState: {
      graphql: {
        introspection: introspectionResponse as unknown as IntrospectionQuery,
        introspectStatus: 'fullfilled',
        apiUrl: 'https://spacex-production.up.railway.app/',
        _persist: { version: -1, rehydrated: true },
      },
    },
  });

  expect(screen.getByText('Documentation')).toBeInTheDocument();
  expect(document.body).toMatchSnapshot();

  await user.click(screen.getByText('Query'));

  expect(screen.getByText('Fields')).toBeInTheDocument();
  const capsuleField = screen.getAllByRole('button', { name: /capsule/ })[0];
  expect(capsuleField).toBeInTheDocument();

  await user.click(capsuleField);

  expect(await screen.findByText('Arguments')).toBeInTheDocument();
});
