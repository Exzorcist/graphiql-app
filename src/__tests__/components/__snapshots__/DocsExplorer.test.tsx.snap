// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DocsEplorer 1`] = `
<body>
  <div>
    <div
      class="h-full flex flex-col relative"
    >
      <div
        class="bg-editor-primary p-5 h-full w-full font-sans min-h-0 flex-[1_1_0px] overflow-auto fancy-scrollbar"
      >
        <h3
          class="text-lg font-semibold"
        >
          Documentation
        </h3>
        <div
          class="shrink-0 bg-editor-border h-[1px] w-full my-2 mb-4"
          data-orientation="horizontal"
          role="none"
        />
        <nav
          aria-label="Breadcrumb"
          class="flex text-sm mb-4"
        >
          <ol
            class="flex flex-wrap items-center gap-x-2"
          >
            <li
              class="inline-flex items-center"
            >
              <button
                class="transition rounded min-w-fit cursor-pointer text-sm py-[2px] px-[3px] hover:bg-transparent hover:text-editor-text-color hover:underline font-semibold"
                role="button"
                tabindex="0"
                type="button"
              >
                Root
              </button>
            </li>
          </ol>
        </nav>
        <div>
          <h4
            class="text-lg font-semibold"
          >
            Root types
          </h4>
          <div
            class="shrink-0 bg-editor-border h-[1px] w-full my-2"
            data-orientation="horizontal"
            role="none"
          />
          <button
            class="py-[5px] px-2 rounded min-w-fit cursor-pointer hover:bg-editor-secondary flex items-center w-full group text-sm hover:text-editor-text-color transition-none"
            role="button"
            tabindex="0"
            type="button"
          >
            query
            : 
            <span
              class="text-editor-accent ml-[6px]"
            >
              Query
            </span>
            <svg
              aria-hidden="true"
              class="text-editor-text-color ml-auto h-4 w-4 opacity-0 group-hover:opacity-100"
              fill="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M12.97 3.97a.75.75 0 011.06 0l7.5 7.5a.75.75 0 010 1.06l-7.5 7.5a.75.75 0 11-1.06-1.06l6.22-6.22H3a.75.75 0 010-1.5h16.19l-6.22-6.22a.75.75 0 010-1.06z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <button
            class="py-[5px] px-2 rounded min-w-fit cursor-pointer hover:bg-editor-secondary flex items-center w-full group text-sm hover:text-editor-text-color transition-none"
            role="button"
            tabindex="0"
            type="button"
          >
            mutation
            : 
            <span
              class="text-editor-accent ml-[6px]"
            >
              Mutation
            </span>
            <svg
              aria-hidden="true"
              class="text-editor-text-color ml-auto h-4 w-4 opacity-0 group-hover:opacity-100"
              fill="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M12.97 3.97a.75.75 0 011.06 0l7.5 7.5a.75.75 0 010 1.06l-7.5 7.5a.75.75 0 11-1.06-1.06l6.22-6.22H3a.75.75 0 010-1.5h16.19l-6.22-6.22a.75.75 0 010-1.06z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <button
            class="py-[5px] px-2 rounded min-w-fit cursor-pointer hover:bg-editor-secondary flex items-center w-full group text-sm hover:text-editor-text-color transition-none"
            role="button"
            tabindex="0"
            type="button"
          >
            subscription
            : 
            <span
              class="text-editor-accent ml-[6px]"
            >
              Subscription
            </span>
            <svg
              aria-hidden="true"
              class="text-editor-text-color ml-auto h-4 w-4 opacity-0 group-hover:opacity-100"
              fill="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M12.97 3.97a.75.75 0 011.06 0l7.5 7.5a.75.75 0 010 1.06l-7.5 7.5a.75.75 0 11-1.06-1.06l6.22-6.22H3a.75.75 0 010-1.5h16.19l-6.22-6.22a.75.75 0 010-1.06z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
